test/unit/math/rev/functor/test_fixture_dae_chem.hpp

```diff
@@ -0,0 +1,398 @@
+
+      auto p1 = theta.at(0);
+      auto p2 = x_r.at(0);
+      auto p3 = x_r.at(1);
+
+      res[0] = yp1 + p1 * yy1 - p2 * yy2 * yy3;
+      res[1] = yp2 - p1 * yy1 + p2 * yy2 * yy3 + p3 * yy2 * yy2;
+      res[2] = yy1 + yy2 + yy3 - 1.0;
+
+      return res;
+    }
+  };
+
+  chemical_kinetics_fun f;
+  chemical_kinetics_data_fun f_data;
+
+  using T_t = std::tuple_element_t<2, T>;
+  using Tyy = std::tuple_element_t<3, T>;
+  using Typ = std::tuple_element_t<4, T>;
+  using T_p = std::tuple_element_t<5, T>;
+
+  std::vector<T_p> theta;
+  Eigen::Matrix<Tyy, -1, 1> yy0;
+  Eigen::Matrix<Typ, -1, 1> yp0;
+  double t0;
+  std::vector<T_t> ts;
+  std::vector<double> x_r;
+  std::vector<int> x_i;
+  double rtol;
+  double atol;
+  int max_num_step;
+
+  chemical_kinetics_dae_base()
+      : theta{0.040, 1.0e4, 3.0e7},
+        yy0(3),
+        yp0(3),
+        t0(0),
+        ts{1, 10, 100, 1000},
+        x_r{1.0e4, 3.0e7},
+        rtol(1.e-5),
+        atol(1.e-12),
+        max_num_step(10000) {
+    yy0 << 1.0, 0.0, 0.0;
+    yp0 << -0.04, 0.04, 0.0;
+  }
+
+  std::vector<T_t>& times() { return ts; }
+  Eigen::VectorXd init() {
+    Eigen::VectorXd joined_init(yy0.size() + yp0.size());
+    joined_init << yy0, yp0;
+    return joined_init;
+  }
+  std::vector<double> param() { return theta; }
+};
+
+/**
+ * Inheriting base type, various fixtures differs by the type of DAE
+ * functor used in <code>apply_solver</code> calls, intended for
+ * different kind of tests.
+ *
+ */
+template <typename T>
+struct chemical_kinetics_test
+    : public chemical_kinetics_dae_base<T>,
+      public ODETestFixture<chemical_kinetics_test<T>> {
+  chemical_kinetics_test() : chemical_kinetics_dae_base<T>() {}
+
+  auto apply_solver() {
+    std::tuple_element_t<0, T> sol;
+    return sol(this->f, this->yy0, this->yp0, this->t0, this->ts, nullptr,
+               this->theta, this->x_r, this->x_i);
+  }
+
+  template <typename T1, typename T2>
+  auto apply_solver(T1&& init, T2&& theta_in) {
+    const int n = init.size() / 2;
+    std::tuple_element_t<0, T> sol;
+    return sol(this->f, init.head(n), init.tail(n), this->t0, this->ts, nullptr,
+               theta_in, this->x_r, this->x_i);
+  }
+
+  auto apply_solver_tol() {
+    std::tuple_element_t<1, T> sol;
+    return sol(this->f, this->yy0, this->yp0, this->t0, this->ts, this->rtol,
+               this->atol, this->max_num_step, nullptr, this->theta, this->x_r,
+               this->x_i);
+  }
+
+  void test_value(double t0_in) {
+    this->t0 = t0_in;
+    for (size_t i = 0; i < this->ts.size(); ++i) {
+      this->ts[i] += t0_in;
+    }
+
+    this->rtol = 1e-4;
+    this->atol = 1e-8;
+    this->max_num_step = 100000;
+    {
+      auto yy = apply_solver_tol();
+      EXPECT_NEAR(0.3368089, stan::math::value_of(yy[3][0]), 1e-4);
+      EXPECT_NEAR(2.01314E-06, stan::math::value_of(yy[3][1]), 1e-9);
+      EXPECT_NEAR(0.6631891, stan::math::value_of(yy[3][2]), 1e-4);
+    }
```

Comment:
How were the tolerances chosen here? Do these correspond approximately to the error between the expected and computed value?

Reply:
Yes. The numeric values are computed from a higher precision (low tolerance).